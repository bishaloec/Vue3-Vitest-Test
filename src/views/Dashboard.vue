<template>
    <div class="dashboard-container">
      <h1>ようこそ、{{ user }} さん！</h1>
      <button @click="handleLogout" class="logout-button">ログアウト</button>
    </div>
  </template>
  
  <script lang="ts">
  import { defineComponent, computed } from 'vue'
  import { useRouter } from 'vue-router'
  import { useAuthStore } from '../stores/auth.ts'
  
  export default defineComponent({
    setup() {
      const authStore = useAuthStore()
      const router = useRouter()
  
      const user = computed(() => authStore.user)
  
      const handleLogout = () => {
        authStore.logout()
        router.push('/login')
      }
  
      return {
        user,
        handleLogout
      }
    }
  })
  </script>
  
  <style scoped>
  .dashboard-container {
    max-width: 800px;
    margin: 50px auto;
    padding: 30px;
    text-align: center;
    position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  }
  
  .logout-button {
    padding: 12px 25px;
    background-color: #dc3545;
    color: white;
    border: none;
    border-radius: 5px;
    cursor: pointer;
    transition: background-color 0.3s;
  }
  
  .logout-button:hover {
    background-color: #c82333;
  }
  </style>