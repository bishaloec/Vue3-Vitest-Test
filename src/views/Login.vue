<template>
  <div class="login-container">
    <h2>ログイン「テスト」</h2>
    
    <div v-if="successMessage" class="success-message">
      {{ successMessage }}
    </div>
    
    <form @submit.prevent="handleLogin" class="login-form">
      <div class="form-group">
        <label for="email">📧 メールアドレス :</label>
        <input 
          id="email"
          v-model="email"
          type="email"
          placeholder="メールアドレスを入力してください"
          required
          @blur="validateEmail"
        />
        <p v-if="emailError" class="error-message">{{ emailError }}</p>
      </div>
      
      <div class="form-group">
        <label for="password">🔒 パスワード:</label>
        <input 
          id="password"
          v-model="password"
          type="password"
          placeholder="パスワードを入力してください"
          required
          @blur="validatePassword"
        />
        <p v-if="passwordError" class="error-message">{{ passwordError }}</p>
        <p v-if="passwordWarning" class="warning-message">{{ passwordWarning }}</p>
      </div>
      
      <div v-if="loginError" class="error-message">
        {{ loginError }}
      </div>
      
      <button 
        type="submit" 
        :disabled="isLoginDisabled || isLoading"
        class="login-button"
      >
        <span v-if="isLoading">ログイン中...</span>
        <span v-else>ログイン</span>
      </button>
    </form>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, computed } from 'vue'
import { useAuthService } from '../services/auth'
import { useAuthStore } from '../stores/auth.ts'
import { useRouter } from 'vue-router'


export default defineComponent({
  name: 'LoginComponent',
  setup() {
    const email = ref('')
    const password = ref('')
    const emailError = ref('')
    const passwordError = ref('')
    const passwordWarning = ref('')
    const loginError = ref('')
    const successMessage = ref('')
    const isLoading = ref(false)
    
    const authService = useAuthService()

    const router = useRouter()
  
    const validateEmail = () => {
      emailError.value = ''
      
      if (!email.value) {
        emailError.value = 'メールアドレスが必要'
        return false
      }
      
      const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
      if (!emailRegex.test(email.value)) {
        emailError.value = '有効なメールアドレスを入力してください'
        return false
      }
      
      return true
    }
    
    const validatePassword = () => {
      passwordError.value = ''
      passwordWarning.value = ''
      
      if (!password.value) {
        passwordError.value = 'パスワードが必要'
        return false
      }
      
      // パスワードの強さを確認
      if (password.value.length < 8) {
        passwordWarning.value = 'パスワードは少なくとも8文字以上の長さが必要です'
      } else if (!/[A-Z]/.test(password.value)) {
        passwordWarning.value = 'パスワードには、少なくとも1つの大文字が含まれている必要があります'
      } else if (!/[0-9]/.test(password.value)) {
        passwordWarning.value = 'パスワードには少なくとも1つの番号が含まれている必要があります'
      }
      
      return true
    }
    
    const isLoginDisabled = computed(() => {
      return !email.value || !password.value || !!emailError.value || !!passwordError.value
    })
    
    const handleLogin = async () => {
      // 以前のエラーをリセット
      loginError.value = ''
      successMessage.value = ''
      
      // フォームを検証
      const isEmailValid = validateEmail()
      const isPasswordValid = validatePassword()
      
      if (!isEmailValid || !isPasswordValid) {
        return
      }
      
      try {
        isLoading.value = true
        const result = await authService.login(email.value, password.value)
        
        if (result.success) {
          successMessage.value = 'ログインが成功'
          const authStore = useAuthStore()
          authStore.login(email.value)
          router.push('/dashboard') // Vue Routerを使用
          // 実際のアプリでは、ここでリダイレクトするか、auth stateを更新する
        } else {
          loginError.value = result.message || 'Login failed. Please check your credentials.'
        }
      } catch (error) {
        loginError.value = 'An unexpected error occurred. Please try again.'
        console.error('Login error:', error)
      } finally {
        isLoading.value = false
      }
    }
    
    return {
      email,
      password,
      emailError,
      passwordError,
      passwordWarning,
      loginError,
      successMessage,
      isLoading,
      isLoginDisabled,
      validateEmail,
      validatePassword,
      handleLogin
    }
  }
})
</script>

<style scoped>
/* Center the login container on the screen */
.login-container {
  max-width: 450px;
  width: 100%;
  padding: 20px;
  background: #fff;
  border-radius: 8px;
  box-shadow: 0px 4px 10px rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.login-form {
  display: flex;
  flex-direction: column;
  gap: 20px;
  width: 100%;
}

.form-group {
  display: flex;
  flex-direction: column;
  gap: 8px;
}

input {
  padding: 10px;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.error-message {
  color: #d9534f;
  font-size: 14px;
  margin-top: 5px;
}

.warning-message {
  color: #f0ad4e;
  font-size: 14px;
  margin-top: 5px;
}

.success-message {
  color: #5cb85c;
  margin-bottom: 20px;
  padding: 10px;
  background-color: #dff0d8;
  border-radius: 4px;
}

.login-button {
  padding: 10px 15px;
  background-color: #337ab7;
  color: white;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}

.login-button:disabled {
  background-color: #cccccc;
  cursor: not-allowed;
}
</style>