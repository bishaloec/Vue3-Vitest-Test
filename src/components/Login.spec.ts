import { describe, it, expect, vi, beforeEach } from 'vitest'
import { mount } from '@vue/test-utils'
import Login from '../views/Login.vue'
import { createRouter, createWebHistory } from 'vue-router'
import { createPinia, setActivePinia } from 'pinia'
import Dashboard from '../views/Dashboard.vue'

// AUTHサービスをMock
vi.mock('@/services/auth', () => ({
  useAuthService: () => ({
    login: vi.fn().mockImplementation(async (email, password) => {
      await new Promise(resolve => setTimeout(resolve, 10))
      
      if (email === 'test@example.com' && password === 'Password123') {
        return { success: true, token: 'mock-token' }
      }
      return { success: false, message: 'ログインメールとパスワードは一致しません' }
    })
  })
}))

// Vue Routerの設定
const router = createRouter({
  history: createWebHistory(),
  routes: [
    { path: '/', redirect: '/login' },
    { path: '/login', component: Login },
    { path: '/dashboard', component: Dashboard, meta: { requiresAuth: true } }
  ]
})

describe('LoginComponent', () => {
  let wrapper: any
  
  beforeEach(async () => {
    // Piniaの初期化
    setActivePinia(createPinia())
    
    // Vue Routerをリセット
    router.push('/login')
    await router.isReady()
    
    // Loginコンポーネントをマウント
    wrapper = mount(Login, {
      global: {
        plugins: [router, createPinia()] // Vue RouterとPiniaをプラグインとして渡す
      }
    })
  })

  it('ログインフォームを正しくレンダリングする', () => {
    expect(wrapper.find('h2').text()).toBe('ログイン「テスト」')
    expect(wrapper.find('input[type="email"]').exists()).toBe(true)
    expect(wrapper.find('input[type="password"]').exists()).toBe(true)
    expect(wrapper.find('button[type="submit"]').exists()).toBe(true)
  })
  
  it('フィールドが空の時にログインボタンを無効にする', async () => {
    const button = wrapper.find('button[type="submit"]')
    expect(button.attributes('disabled')).toBeDefined()
    
    // メールのみに記入
    await wrapper.find('input[type="email"]').setValue('test@xyz.com')
    expect(button.attributes('disabled')).toBeDefined()
    
    // パスワードのみに記入
    await wrapper.find('input[type="email"]').setValue('')
    await wrapper.find('input[type="password"]').setValue('P@ssword123')
    expect(button.attributes('disabled')).toBeDefined()
    
    // 両方のフィールドに入力
    await wrapper.find('input[type="email"]').setValue('test@example.com')
    await wrapper.find('input[type="password"]').setValue('P@ssword123')
    expect(button.attributes('disabled')).toBeUndefined()
  })
  
  it('メール形式を検証する', async () => {
    const emailInput = wrapper.find('input[type="email"]')
    
    // 無効な電子メールをテスト
    await emailInput.setValue('invalid-email')
    await emailInput.trigger('blur')
    expect(wrapper.find('.error-message').text()).toContain('有効なメールアドレスを入力')
    
    // 有効な電子メールをテスト
    await emailInput.setValue('test@xyz.com')
    await emailInput.trigger('blur')
    expect(wrapper.find('.error-message').exists()).toBe(false)
  })
  
  it('パスワード筋力警告を表示する', async () => {
    const passwordInput = wrapper.find('input[type="password"]')
    
    // 短いパスワードをテスト
    await passwordInput.setValue('Short')
    await passwordInput.trigger('blur')
    expect(wrapper.find('.warning-message').text()).toContain('8文字以上の長さ')
    
    // 大文字のないパスワードをテスト
    await passwordInput.setValue('password123')
    await passwordInput.trigger('blur')
    expect(wrapper.find('.warning-message').text()).toContain('1つの大文字が含まれている必要')
    
    // 数字なしでパスワードをテスト
    await passwordInput.setValue('Password')
    await passwordInput.trigger('blur')
    expect(wrapper.find('.warning-message').text()).toContain('1つの番号が含まれている必要')
    
    // 強力なパスワードをテスト
    await passwordInput.setValue('Password123')
    await passwordInput.trigger('blur')
    expect(wrapper.find('.warning-message').exists()).toBe(false)
  })
  
  it('成功したログインを処理する', async () => {
    // 有効なメールとパスワードでフォームに記入
    await wrapper.find('input[type="email"]').setValue('test@example.com')
    await wrapper.find('input[type="password"]').setValue('Password123')
    
    // フォームをsubmit
    await wrapper.find('form').trigger('submit')
    
    // 読み込み状態を確認
    expect(wrapper.find('button').text()).toContain('ログイン中')
    
    // モックAPI呼び出しが解決するのを待ち
    await new Promise(resolve => setTimeout(resolve, 20))
    
    // ルーターがダッシュボードに移動するかどうかを確認
    expect(router.currentRoute.value.path).toBe('/dashboard')
  })
  
  it('失敗したログインにエラーメッセージを表示する', async () => {
    // 無効なメールとパスワードでフォームに記入
    await wrapper.find('input[type="email"]').setValue('wrong@example.com')
    await wrapper.find('input[type="password"]').setValue('WrongPass123')
    
    // フォームをsubmit
    await wrapper.find('form').trigger('submit')
    
    // モックAPI呼び出しが解決するのを待ち
    await new Promise(resolve => setTimeout(resolve, 20))
    
    // エーラーメッセージをチェック
    expect(wrapper.find('.error-message').text()).toEqual('ログインメールとパスワードは一致しません')
  })

  it('ログイン snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
})