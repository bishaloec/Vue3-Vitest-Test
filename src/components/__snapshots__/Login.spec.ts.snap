// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginComponent > ログイン snapshot 1`] = `
"<div data-v-08aaad4e="" class="login-container">
  <h2 data-v-08aaad4e="">ログイン「テスト」</h2>
  <!--v-if-->
  <form data-v-08aaad4e="" class="login-form">
    <div data-v-08aaad4e="" class="form-group"><label data-v-08aaad4e="" for="email">📧 メールアドレス :</label><input data-v-08aaad4e="" id="email" type="email" placeholder="メールアドレスを入力してください" required="">
      <!--v-if-->
    </div>
    <div data-v-08aaad4e="" class="form-group"><label data-v-08aaad4e="" for="password">🔒 パスワード:</label><input data-v-08aaad4e="" id="password" type="password" placeholder="パスワードを入力してください" required="">
      <!--v-if-->
      <!--v-if-->
    </div>
    <!--v-if--><button data-v-08aaad4e="" type="submit" disabled="" class="login-button"><span data-v-08aaad4e="">ログイン</span></button>
  </form>
</div>"
`;

exports[`LoginComponent > 失敗したログインにエラーメッセージを表示する 1`] = `
"<div data-v-08aaad4e="" class="login-container">
  <h2 data-v-08aaad4e="">ログイン「テスト」</h2>
  <!--v-if-->
  <form data-v-08aaad4e="" class="login-form">
    <div data-v-08aaad4e="" class="form-group"><label data-v-08aaad4e="" for="email">📧 メールアドレス :</label><input data-v-08aaad4e="" id="email" type="email" placeholder="メールアドレスを入力してください" required="">
      <!--v-if-->
    </div>
    <div data-v-08aaad4e="" class="form-group"><label data-v-08aaad4e="" for="password">🔒 パスワード:</label><input data-v-08aaad4e="" id="password" type="password" placeholder="パスワードを入力してください" required="">
      <!--v-if-->
      <!--v-if-->
    </div>
    <div data-v-08aaad4e="" class="error-message">ログインメールとパスワードは一致しません</div><button data-v-08aaad4e="" type="submit" class="login-button"><span data-v-08aaad4e="">ログイン</span></button>
  </form>
</div>"
`;
